doc/developer/design/20240108_source_metrics_2.md

```diff
@@ -0,0 +1,231 @@
+f2: {0: 1010th gtid, 1: 10th gtid}
+```
+In this case, `f2 - f1 = 20 transactions`
+
+#### Comparisons
+
+As you can see, source frontier differences depend on the source, and are primarily useful for _calculating rates_ and when
+_comparing values_. For example:
+
+- If a source's frontier has moved from `{0: 10}` to `{0: 70}` in 1 minute, Materialize can be said to have processed 1 offset/minute.
+- If a source's frontier snapshot frontier is `{table1: 100 rows}` and its current frontier is `{table1: 50}`, Materialize can be said
+to have processed about 50% of the snapshot.
+
+Additionally, differences are designed to be _lower-bounded estimates_. For example, `{0: 100} - {0: 0, 1: 50}` is 100, as we
+choose the lower bound where we assume no offsets past 50 for partition 1 exist.
+
+
+## Snapshot Progress
+
+The first set of metrics this design document proposes involve _snapshot progress_, which is a _lower bound_ on the _percentage_ of the source's snapshot Materialize
+has _read_. These metrics are designed to answer #1 in [the problem statement](#the-problem).
+
+We will introduce 2 new columns in `mz_source_statistics_per_worker`:
+
+```
+| `snapshot_total`     | [`uint8`] | The total number of upstream values that are part of the snapshot. |
+| `snapshot_progress   | [`uint8`] | The number of upstream values Materialize has read so far.         |
+```
+
+The unit of _values_ depends on the source type, and will be _rows_ for MySQL and Postgres, and _offsets_ for kafka.
+
+These values can be summed across workers and compared (`snapshot_progress / snapshot_progress - snapshot_total`) to produce
+a _lower-bound_ estimate on the progress we have made reading the source's snapshot.
+
+### Source frontier difference
+
+Note that this percentage is calculated using _source frontier difference_, between the `snapshot_total` and `snapshot_progress`.
+For consistency within schema of the `mz_source_statistics_per_worker` table, each worker in the source pipeline is required to
+pre-aggregate these values from the possibly complex frontier that represents the total and progress of a snapshot.
+This primarily means summing offsets/rows across partitions/tables.
+
+### Source specifics
+
+The `SourceReader` implementation for each source will be required to produce a `snapshot_total`, as well as
+a continuously updated `snapshot_progress` frontier on each worker.
+
+#### Kafka
+
+`snapshot_total` can be be trivially exposed by exposing the snapshot frontier already tracked within its source operator,
+and summing across partitions. Similarly, `snapshot_progress` can be derived from the operator's output frontier.
+
+#### Postgres and MySQL
+
+`snapshot_total` will need to be calculated, in the unit of rows by performing `SELECT count(*)` on the tables that participate in the snapshot.
+Both the Postgres and MySQL implementations will be required to perform this query, per-table, during snapshotting. For RDBMSes, this
+should be cheap.
```

Comment:
Wait, `SELECT count(*)` is generally decidedly not cheap for RDBMSes: https://wiki.postgresql.org/wiki/Slow_Counting

Perhaps it's fine because we're going to `SELECT *` anyway, but it's definitely not free, and we should think twice about the implications of doing a count(*) first. It might add meaningful overhead to the snapshot time.

Reply:
(@rjobanp you were also interested in this)

@benesch, I did some research and testing, and while mysql and postgres both have count estimates, they both are problematic; Mysql's can be (as the documentation suggests) 40-50% wrong in innodb, and postgres returns -1 for tables that haven't been `ANALYZED` or vacuumed recently :(

I think we should go ahead and start with `count(*)` over an estimate, but critically: run this count concurrently with the snapshot being processed (I have a local branch that is almost ready that does this, but also offers LD configuration for all these options). This will allow us to at least report the _number of rows we have processed_ before we get a full count, which can still be useful for the user. Ill update the doc to reflect this.

The local branch also lets see in production what real world `count(*)`'s really cost!

Reply:
That seems reasonable! In PostgreSQL, I think you might be able to use `pg_export_snapshot()` and `SET TRANSACTION SNAPSHOT` to allow the second connection you create to count the tables to get the exact same number of rows that the snapshot connection will eventually read.

Reply:
Yes, thats what I intend to do!